flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/DynamicKafkaRecordSerializationSchema.java

```diff
@@ -74,25 +113,63 @@ class DynamicKafkaRecordSerializationSchema implements KafkaRecordSerializationS
     @Override
     public ProducerRecord<byte[], byte[]> serialize(
             RowData consumedRow, KafkaSinkContext context, Long timestamp) {
+
+        // keeping the metadataHeaders maps for keys and values separate so these maps
+        // are output only for the methods; avoiding changing the inputs.
+        // define the input map
+        Map<String, Object> inputMap = new HashMap<>();
+
+        // define the output maps, it is 2 levels deep in case we need to add new information in
+        // the future
+        Map<String, Object> outputKeyMap = new HashMap<>();
+        Map<String, Object> outputValueMap = new HashMap<>();
+
+        inputMap.put(TOPIC_NAME, topic);
+
         // shortcut in case no input projection is required
         if (keySerialization == null && !hasMetadata) {
-            final byte[] valueSerialized = valueSerialization.serialize(consumedRow);
-            return new ProducerRecord<>(
-                    topic,
-                    extractPartition(
-                            consumedRow,
-                            null,
-                            valueSerialized,
-                            context.getPartitionsForTopic(topic)),
-                    null,
-                    valueSerialized);
+            byte[] valueSerialized = getSerialized(consumedRow, inputMap, outputValueMap, false);
+            if (serializeWithAdditionalPropertiesMethod == null) {
```

Comment:
This logic seems odd - you check for `serializeWithAdditionalPropertiesMethod` but then you do not use it in the `else` block.

It could be re-written to be cleaner regardless - the logic is only about adding headers, so you can put the header generation in an if  block and then create ProducerRecord later in both cases.